RFC-0011-tvf.md

```diff
@@ -0,0 +1,752 @@
+
+```java
+public class DescriptorArgumentSpecification
+       extends ArgumentSpecification
+{
+   private DescriptorArgumentSpecification(String name, boolean required, Descriptor defaultValue)
+   {
+       super(name, required, defaultValue);
+   }
+} 
+
+public class Descriptor
+{
+   private final List<Field> fields; 
+   
+   @JsonCreator
+   public Descriptor(@JsonProperty("fields") List<Field> fields)
+   {
+       requireNonNull(fields, "fields is null");
+       checkArgument(!fields.isEmpty(), "descriptor has no fields");
+       this.fields = unmodifiableList(fields);
+   } 
+   
+   public static Descriptor descriptor(String... names)
+   {
+       List<Field> fields = Arrays.stream(names)
+               .map(name -> new Field(name, Optional.empty()))
+               .collect(Collectors.toList());
+       return new Descriptor(fields);
+   } 
+   
+   public static Descriptor descriptor(List<String> names, List<Type> types)
+   {
+       requireNonNull(names, "names is null");
+       requireNonNull(types, "types is null");
+       checkArgument(names.size() == types.size(), "names and types lists do not match");
+       List<Field> fields = new ArrayList<>();
+       for (int i = 0; i < names.size(); i++) {
+           fields.add(new Field(names.get(i), Optional.of(types.get(i))));
+       }
+       return new Descriptor(fields);
+   } 
+   
+   public static class Field
+   {
+       private final String name;
+       private final Optional<Type> type; 
+       
+       @JsonCreator
+       public Field(@JsonProperty("name") String name, @JsonProperty("type") Optional<Type> type)
+       {
+           this.name = checkNotNullOrEmpty(name, "name");
+           this.type = requireNonNull(type, "type is null");
+       }
+   }
+}
+```
+
+#### Table argument
+
+This represents a table input. Table arguments have some additional properties.
+
+##### Row or Set semantics
+
+A table argument with set semantics is always invoked with a PARTITION BY clause and the table function is processed on
+a partition by partition basis. If no partitioning on a table argument with set semantics then the entire input is
+processed as a single partition.
+
+A table argument with row semantics is processed on a row-by-row basis. Partitioning or ordering is not applicable.
+
+##### Prune or keep when empty
+
+Prune when empty property indicates that if the table argument is empty then the function returns empty result. 
+
+If keep when empty then the function is executed even if the table argument is empty. It is similar to the inner/outer join property.
+
+##### Pass-through columns
+
+All columns of the table are pass-through if this property is set. Else only partition columns of the table are pass-through.
+
+```java
+public class TableArgumentSpecification
+       extends ArgumentSpecification
+{
+   private final boolean rowSemantics;
+   private final boolean pruneWhenEmpty; 
+   private final boolean passThroughColumns; 
```

Comment:
While I can't think of a way to simplify this, this seems complicated to debug if the connector author doesn't get it right.  I'm wondering what sort of testing tools we can give to connector authors to validate the behavior of their functions?  For example, is it possible to see the plan of a proposed table valued function and validate that the plan is correct?  It would seem like any TVF would want to write tests that validate the TVF is correctly planned, right?

Reply:
Agree. It might be complicated for function authors to test different combinations of SQL to ensure they are planned and evaluated correctly. We might need to develop a fuzzer of sorts for it. I'll put some thought into how we can do it in a principled way.